File: J9Ch/src/J_5_Depth_First_Search/Required_7/_153_Combination_Sum_II_Medium.java
package J_5_Depth_First_Search.Required_7;
import java.util.*;import lib.TreeNode;import lib.AssortedMethods;import org.junit.Test;
/** 153. Combination Sum II
 * Medium

 * Created by tianhuizhu on 6/27/17.
 */
public class _153_Combination_Sum_II_Medium {

    public class Solution {
        /**
         * @param num: Given the candidate numbers
         * @param target: Given the target number
         * @return: All the combinations that sum to target
         */
        public List<List<Integer>> combinationSum2(int[] candidates,
                                                   int target) {
            List<List<Integer>> results = new ArrayList<>();
            if (candidates == null || candidates.length == 0) {
                return results;
            }

            Arrays.sort(candidates);
            List<Integer> combination = new ArrayList<Integer>();
            helper(candidates, 0, combination, target, results);

            return results;
        }

        private void helper(int[] candidates,
                            int startIndex,
                            List<Integer> combination,
                            int target,
                            List<List<Integer>> results) {
            if (target == 0) {
                results.add(new ArrayList<Integer>(combination));
                return;
            }

            for (int i = startIndex; i < candidates.length; i++) {
                if (i != startIndex && candidates[i] == candidates[i - 1]) {
                    continue;
                }
                if (target < candidates[i]) {
                    break;
                }
                combination.add(candidates[i]);
                helper(candidates, i + 1, combination, target - candidates[i], results);
                combination.remove(combination.size() - 1);
            }
        }
    }
}

File: J9Ch/src/J_5_DFS/Required_7/python_cpp/_153_Combination_Sum_II_Medium.cpp
/** 153. Combination Sum II
 * Medium*/


class Solution {
public:
	/**
	 * @param num: Given the candidate numbers
	 * @param target: Given the target number
	 * @return: All the combinations that sum to target
	 */
    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
		// write your code here
        sort(num.begin(), num.end());
        vector<int> stack, rec; int sum=0, cur=0;
        vector<vector<int> > ans;
        while (cur<num.size() && num[cur]+sum<=target) {
			stack.push_back(num[cur]);
			rec.push_back(cur);
			sum+=num[cur++];
		}
		if (sum==target) ans.push_back(stack);
		while (rec.size()!=0) {
			cur=rec.back(); int x=num[cur];
			sum-=x; stack.pop_back(); rec.pop_back();
			for (++cur; cur<num.size(); ++cur)
				if (num[cur]!=x || sum+num[cur]>target) break;
			if (cur<num.size() && num[cur]!=x)
				while (cur<num.size() && num[cur]+sum<=target) {
					stack.push_back(num[cur]);
					rec.push_back(cur);
					sum+=num[cur++];
				}
			if (sum==target) ans.push_back(stack);
		}
		return ans;
    }
};